srv/src/main/java/my/bookshop/handlers/CatalogServiceHandler.java

```diff
@@ -2,56 +2,223 @@
+		books.filter(b -> b.getTitle() != null).forEach(b -> {
+			loadStockIfNotSet(b);
+			discountBooksWithMoreThan111Stock(b);
+		});
+	}
+
+	@After
+	public void setIsReviewable(CdsReadEventContext context, List<Books> books) {
+		String user = context.getUserInfo().getName();
+
+		Result result = db
+				.run(Select.from(CatalogService_.REVIEWS).columns(Reviews_::book_ID).where(r -> r.reviewer().eq(user)));
+		Set<String> reviewedBooks = result.streamOf(Reviews.class).map(Reviews::getBookId).collect(Collectors.toSet());
+
+		for (Books book : books) {
+			if (reviewedBooks.contains(book.getId())) {
+				book.setIsReviewable(false);
 			}
-		);
+		}
+	}
+
+	@On
+	public void onSubmitOrder(SubmitOrderContext context) {
+		Integer amount = context.getAmount();
+		String bookId = context.getBook();
+
+		Optional<Books> book = db.run(Select.from(BOOKS).columns(Books_::stock).byId(bookId)).first(Books.class);
+
+		book.orElseThrow(() -> new ServiceException(ErrorStatuses.NOT_FOUND, MessageKeys.BOOK_MISSING)
+				.messageTarget(Books_.class, b -> b.ID()));
+
+		Integer stock = book.map(Books::getStock).get();
+
+		if (stock >= amount) {
+			db.run(Update.entity(BOOKS).byId(bookId).data(Books.STOCK, stock -= amount));
+
+			ReturnCatalogServiceSubmitOrder result = ReturnCatalogServiceSubmitOrder.create();
+			result.setStock(stock);
+
+			context.setResult(result);
+			context.setCompleted();
+		} else {
+			throw new ServiceException(ErrorStatuses.CONFLICT, String.format("%d exceeds stock for book", amount));
+		}
 	}
 
 	private void discountBooksWithMoreThan111Stock(Books b) {
-		if(b.getStock() > 111) {
+		if (b.getStock() != null && b.getStock() > 111) {
 			b.setTitle(String.format("%s -- 11%% discount", b.getTitle()));
 		}
 	}
 
 	private void loadStockIfNotSet(Books b) {
-		if (b.getStock() == null) {
-			b.setStock(db.run(Select.from(BOOKS).byId(b.getId()).columns(Books_::stock)).single(Books.class).getStock());
+		if (b.getId() != null && b.getStock() == null) {
+			b.setStock(
+					db.run(Select.from(BOOKS).byId(b.getId()).columns(Books_::stock)).single(Books.class).getStock());
+		}
+	}
+
+	private void validateReview(Reviews review) {
+		validateBook(review);
+		validateRating(review);
+	}
+
+	private void validateRating(Reviews review) {
+		Integer rating = review.getRating();
+		if (rating == null) {
+			long r = Math.round(Math.random() * 4) + 1;
+			review.setRating(Integer.valueOf((int) r));
+		} else {
+			if (rating < 1 || rating > 5) {
+				throw new ServiceException(ErrorStatuses.BAD_REQUEST, MessageKeys.REVIEW_INVALID_RATING)
+						.messageTarget(Reviews_.class, r -> r.rating());
+			}
 		}
 	}
+
+	private void validateBook(Reviews review) {
+		if (review.getBookId() == null) {
+			throw new ServiceException(ErrorStatuses.BAD_REQUEST, MessageKeys.BOOK_MISSING)
+					.messageTarget(Reviews_.class, r -> r.book_ID());
+		}
+
+		Optional<Row> row = db.run(Select.from(CatalogService_.BOOKS).byId(review.getBookId())).first();
```

Comment:
Why do you use the `PersistenceService` and not the `CatalogService` here?

Reply:
Primarily, because the `PersistenceService` injection existed before already and I decided to reuse it instead of injecting the `CatalogService` in addition. Is it preferred to use the `CatalogService`?